### ISSUE TITLE
more examples with real numbers in README

### ISSUE THREAD

**Issue by lapsio** [markfasheh/duperemove#148]
Would it be possible to add more examples with real data and requirements? For example I'd like to run dedup on machine with 16gb ram on 5TB RAID1 and 6TB RAID6. What is estimated memory requirement? should i use hashfile? What would be ETA hashfile size and ETA memory usage without hashfile?

Maybe even some rules of thumb how to estimate required hashfile size or required ram with various options? It'd be really helpful for "preparation" before running dedup on big containers

**Comment by markfasheh** [comment]
Yeah that's something I'm working on actually. There's a couple pages in the wiki already, and I intend to update the design doc one this week. It includes resource usage for each stage, however I believe the text could be more specific (so I'll try to add that).

https://github.com/markfasheh/duperemove/wiki/Duperemove-Design

**Comment by markfasheh** [comment]
Btw, to actually answer the question (I'll put most of this on the wiki too):


The csum stage is going to take a bit of cpu and as much disk bandwidth as you'll give it. Memory is all going to be in buffers - we're only streaming to the hashfile at this point. Reads are entirely sequential.


The most resource intensive stage is the 'Find-dupes phase' where we build a list of extents that are duplicated amongst the data set. In a way it's re-assembling blocks back into extents. You can actually skip this stage with '--dedupe-options=block' if you like and duperemove will just do dedupe on 128K extents (or whatever blocksize you specify).

Resource usage of this stage is proportional to the amount of duplicated data we have to process, so total amount of input data doesn't help (though if we know a "% duplicated", we can calculate from there). Mostly though the usage is in two structures - an 'extent' structure which does the tracking and 2ndly  a 'compares' tree we keep on each file to avoid running the same extents more than once.

Most of the memory this stage uses is freed afterwards.


The dedupe stage is similar to the csum stage - we're running our results tree and deduping. We'll be doing mostly I/O here (data reads and metadata updates) and calls to the fiemap ioctl (disable this with --dedupe-options=nofiemap). So resource usage here is mostly disk i/o unless you have an expensive fiemap (btrfs currently does but this is being worked on).

**Comment by lapsio** [comment]
That answer is quite far from real actual numbers :D Is 20gb on SSD enough for hashfile? Or should I get more? 30? 40? 100? 500? I think there can be a lot of duplicated data that's why I'd like to dedup. I can roughly estimate out of those 6TB probably about 1TB or even more might be duplicated, including few hundred gb of huge areas with zero filled binary files. Really "dedup-friendly" data.

**Comment by lapsio** [comment]
Also if I dedup with bigger extents first, then with smaller ones will that be better / worse than doing with smaller ones from very beginning? I mean both dedup quality and resource requirements wise

**Comment by markfasheh** [comment]
Ok thanks for looking that over. I'll get us some real examples. On the topic of hashfile size, here's the hashfile for my home directory:

    $ ls -alh slash-home-scanonly.dup 
    -rw-r--r-- 1 mfasheh users 500M Sep 15 18:53 slash-home-scanonly.dup

Which `df` reports as being `727G` in size (but not all of it is stored in the hashfile).

Hashfile size will depend on the # of files + the # of hashes stored. Those are the only two objects stored in the database.

Here are the row counts of each table:

    sqlite> select count(*) from files;
    128724
    sqlite> select count(*) from hashes;
    5849382

the hashes are 128k each, so `5849382*128K` gives us about `714G` of data represented in this file.

`sqlite3_analyzer` tells us how big each table is (output edited for brevity):

    Page size in bytes................................ 1024

    *** Page counts for all tables with their indices *****************************

    HASHES............................................ 477285      93.4% 
    FILES............................................. 33979        6.6% 
    CONFIG............................................ 2            0.0% 
    SQLITE_MASTER..................................... 1            0.0% 
    
So for hashes table, each entry takes about 83 bytes:

    (477285 * 1024)/5849382
    83

And for files table, each entry takes about 270 bytes (this will vary based on filename):

    (33979*1024)/128724
    270


Now if I've done everything right, the formula for hashfile size should be (minus overhead):

**Hashfile Size** = **Num Hashes** X **83**  +  **Num Files** X **270**

so if we plug in for your case i get `6597069766656 bytes / 83` which gives us about 74 GB, not including file overhead.


Please check my math :)

**Comment by markfasheh** [comment]
Regarding your 2nd question,

If the total # of duplicated blocks is the same between data sets resource usage of duperemove will be higher for the one with **more** duplicated extents since we'll have more memory for tracking them, and more dedupe requests to send down.

Dedupe quality wise, you might want to try deduping the files with bigger extents first. It's hard to say without knowing how much both sets overlap but it's probably better to have the files with smaller extents change their references to the larger extents. In the end though I'm not really sure how much of a difference it will make.

**Comment by ian-kelling** [comment]
+1 to this request. I ran duperemove on 2 snapshots which were mostly duplicate, about 500G combined, about 20k files, called with --dedupe-options=nofiemap  --hashfile=qduperemove -dhrA snapshot1 snapshot2. I stopped it after 4 days of being in the Find-dupes phase. Top said 500% cpu every time I checked it. It looks like there is a % complete bar, it hadn't changed at all. I wonder how long it would have taken? I'm rerunning with --dedupe-options=block, which is in progress now. Since it's 2 snapshots with mostly common directories and filenames, it would have been a good idea to split the job up into running on subsets of the snapshots. Splitting up a big job if possible is probably good advice to put in the documentation.

**Comment by markfasheh** [comment]
> +1 to this request. I ran duperemove on 2 snapshots which were mostly duplicate, about 500G combined, about 20k files, called with --dedupe-options=nofiemap --hashfile=qduperemove -dhrA snapshot1 snapshot2. I stopped it after 4 days of being in the Find-dupes phase. Top said 500% cpu every time I checked it

Hi Ian, is that hashfile something you'd be willing to share with me? My data sets are mostly synthetic (see https://github.com/markfasheh/duperemove-tests/blob/master/src/gen_mkzeros_test_data.sh for example) so I'm basically discovered performance corner cases as I can come up with them. Any real world data would help a ton.

> It looks like there is a % complete bar, it hadn't changed at all. I wonder how long it would have taken?

A long time by the sound of it, sorry :/   I'm very glad you were able to get dedupe with --dedupe-options=block. This was exactly the kind of case I intended for it - I really hate to hear that a user gets stuck like you did. Estimating the actual time to completion is difficult but what that bar does is tell you how many (out of 100%) of the dup buckets we've visited. So probably you had an enormous bucket with a ton of files and we were still comparing those. If you run with '-v' it will give you more info on what's going on.

So find_dupes is probably the last part of duperemove that I feel is still kind of rough. The objective is to optimize our dedupe requests to cut down on the number of extent references btrfs creates - Btrfs in general is poor at merging on-disk extent references. The algorithm is pretty gross, performance-wise. We don't get a lot of help here and have to reassemble blocks into extents. I don't actually know of any other software that does this (what that tells us is up to interpretation ;).


So it sounds like we could be more clear in the man page about two additional things:

-  the find dupes step and what to expect, and that  if it goes AWOL you can simply restart the same command with --dedupe-options=block .

- And of course, noting that it might be best to break up your data set into smaller ones and dedupe those

**Comment by markfasheh** [comment]
Hey along the lines of telling the user what's going on and how to avoid some bad situations, maybe we should alter the print that happens before find_dupes:

"Hashing completed. Using 4 threads to calculate duplicate extents. This may take some time. If  performance during this stage is deemed to be unacceptable, you may interrupt the program and restart with the  '--dedupe-options=block'  command line option"

**Comment by ian-kelling** [comment]
> "Hashing completed. Using 4 threads to calculate duplicate extents. This may take some time. If performance during this stage is deemed to be unacceptable, you may interrupt the program and restart with the '--dedupe-options=block' command line option"

Yes

**Comment by ian-kelling** [comment]
> So it sounds like we could be more clear in the man page about two additional things:
> 
>     the find dupes step and what to expect, and that if it goes AWOL you can simply restart the same command with --dedupe-options=block .
> 
>     And of course, noting that it might be best to break up your data set into smaller ones and dedupe those

Yes

> Hi Ian, is that hashfile something you'd be willing to share with me?

I'm assuming it contains filenames in it, in this case I can't. I will let you know if I come across a very bad performing one which I can.

**Comment by markfasheh** [comment]
Yeah it'll have filenames in it though i suppose you could obfuscate though with enough fancy SQL  :)

Actually if you give me the output of hashstats on one of those that might work too -- it'll give me a file and block count for the top buckets but should not print anything identifiable.

**Comment by markfasheh** [comment]
Ok, I made several changes to the man page in master branch. In particular, we have a new FAQ section on estiming the size of a hashfile. I also added information on memory usage and breaking up the dataset in the first FAQ ("Is there an upper limit to the amount of data duperemove can process?")

In addition, block dedupe is the default now and I note in the man page that if the find-dupes step is run it will consume loads of memory.
